### ISSUE TITLE
DDP training hangs when one GPU returns zero loss

### ISSUE THREAD

**Issue by kazimpal87** [pytorch/pytorch#73645]
Hi

This issue is similar to https://github.com/pytorch/pytorch/issues/23425

I have a situation where I am training a multitask detection model. The model is performing object detection as well as a number of subsequent classification tasks on each detected bounding box, but the loss is computed only on these classification tasks. The detection part of the model is frozen. 

This means that if a batch of images contains no ground truth objects (or there are ground truth objects but they just arent detected), it is effectively a batch of size zero, and the loss should be zero. In a single-gpu context I get away with just creating a zero tensor as the loss, but this doesn't work in the DDP context and causes the training to hang. I presume this is to do with the fact that the loss needs to contain gradients to be gathered and reduced across the gpus.

Is there a way to solve this problem? Say by creating a zero-tensor with all of the appropriate gradients present?

### Versions

PyTorch version: 1.10.2+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.5 (default, Jul 28 2020, 12:59:40)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-96-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2080
GPU 1: NVIDIA GeForce RTX 2080
GPU 2: NVIDIA GeForce RTX 2080
GPU 3: NVIDIA GeForce RTX 2080

Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] facenet-pytorch==2.5.2
[pip3] numpy==1.22.2
[pip3] pytorch-lightning==1.5.10
[pip3] pytorch-metric-learning==1.2.0
[pip3] torch==1.10.2+cu113
[pip3] torchaudio==0.10.2+cu113
[pip3] torchmetrics==0.7.2
[pip3] torchvision==0.11.3+cu113
[conda] Could not collect

**Comment by KrishPro** [comment]
Why don't you filter all the data points without ground truth, Before Hand ?

**Comment by dagitses** [comment]
Hi, We use github issues only for bug reports or feature requests. Please use the forum to ask questions: https://discuss.pytorch.org/

**Comment by KrishPro** [comment]
I would suggest mentioning the link to your forum page here, So we can help you there !
